File: January 2025/31.cpp
class Solution {
public:
    // Function to find a solved Sudoku.
    void solveSudoku(std::vector<std::vector<int>>& mat) {
        std::vector<int> r(9, 0), c(9, 0), box(9, 0);
        
        // Initialize the state of rows, columns, and boxes
        for (int i = 0; i < 9; ++i) {
            for (int j = 0; j < 9; ++j) {
                if (mat[i][j] != 0) {
                    int m = 1 << mat[i][j], idx = i / 3 * 3 + j / 3;
                    r[i] |= m; 
                    c[j] |= m; 
                    box[idx] |= m;
                }
            }
        }

        solve(mat, r, c, box, 0, 0);
    }

private:
    bool solve(std::vector<std::vector<int>>& b, std::vector<int>& r, std::vector<int>& c, std::vector<int>& box, int i, int j) {
        if (i == 9) return true;
        if (j == 9) return solve(b, r, c, box, i + 1, 0);
        if (b[i][j] != 0) return solve(b, r, c, box, i, j + 1);

        for (int num = 1; num <= 9; ++num) {
            int m = 1 << num;
            int idx = i / 3 * 3 + j / 3;
            if ((r[i] & m) || (c[j] & m) || (box[idx] & m)) continue;
            b[i][j] = num;
            r[i] |= m;
            c[j] |= m;
            box[idx] |= m;
            if (solve(b, r, c, box, i, j + 1)) return true;
            b[i][j] = 0;
            r[i] &= ~m;
            c[j] &= ~m;
            box[idx] &= ~m;
        }

        return false;
    }
};

File: January 2025/31.py
class Solution:

    def solveSudoku(self, mat):
        def solve(b, r, c, box, i, j):
            if i == 9:
                return True
            if j == 9:
                return solve(b, r, c, box, i + 1, 0)
            if b[i][j] != 0:
                return solve(b, r, c, box, i, j + 1)
            for num in range(1, 10):
                m = 1 << num
                idx = i // 3 * 3 + j // 3
                if (r[i] & m) != 0 or (c[j] & m) != 0 or (box[idx] & m) != 0:
                    continue
                b[i][j] = num
                r[i] |= m
                c[j] |= m
                box[idx] |= m
                if solve(b, r, c, box, i, j + 1):
                    return True
                b[i][j] = 0
                r[i] &= ~m
                c[j] &= ~m
                box[idx] &= ~m
            return False

        r = [0] * 9
        c = [0] * 9
        box = [0] * 9
        for i in range(9):
            for j in range(9):
                if mat[i][j] != 0:
                    m = 1 << mat[i][j]
                    idx = i // 3 * 3 + j // 3
                    r[i] |= m
                    c[j] |= m
                    box[idx] |= m
        solve(mat, r, c, box, 0, 0)
